### PR TITLE
Tree printer for Python AST and its ASR

### PR DESCRIPTION
**Pull request by ansharlubis**
- Implemented a tree printer for Python AST by adding a class generating visitor `TreeVisitorVisitor` in `asdl_cpp.py`.
- To execute the printing, I added the option `--tree` during AST or ASR printing.
- New functions are added to `pickle.h` to handle the printing.

### CODE REVIEW

**Comment by certik** [comment]
Awesome, thanks a lot of this!

The Python script seems to fail on Windows at the CI with the error:
```
Traceback (most recent call last):
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 1568, in <module>
    sys.exit(main(sys.argv))
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 1560, in main
    visitor(fp, data).visit(mod)
  File "D:\a\lpython\lpython\grammar\asdl.py", line 128, in visit
    meth(obj, *args)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 442, in visitModule
    super(TreeVisitorVisitor, self).visitModule(mod)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 19, in visitModule
    self.visit(df, *args)
  File "D:\a\lpython\lpython\grammar\asdl.py", line 128, in visit
    meth(obj, *args)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 446, in visitType
    super(TreeVisitorVisitor, self).visitType(tp, tp.name)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 26, in visitType
    self.visit(tp.value, *args)
  File "D:\a\lpython\lpython\grammar\asdl.py", line 128, in visit
    meth(obj, *args)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 455, in visitSum
    self.visit(tp, *args)
  File "D:\a\lpython\lpython\grammar\asdl.py", line 128, in visit
    meth(obj, *args)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 461, in visitConstructor
    self.make_visitor(cons.name, cons.fields, True)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 472, in make_visitor
    self.emit(              'last ? s.append("\u2514-") : s.append("|-");', 3)
  File "D:\a\lpython\lpython\grammar\asdl_cpp.py", line 41, in emit
    self.stream.write(indent + line + "\n")
  File "C:\Miniconda\envs\test\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\u2514' in position 29: character maps to <undefined>
```
I think we have to make sure the string is in UTF-8, and then encode the character properly, I can't remember right now how it is done in Python, but there is a way to do it, so that it works on all platforms.

**Comment by ansharlubis** [comment]
I just made another commit specifying the encoding to "UTF-8", do I need to do anything to rerun the CI?

**Comment by certik** [comment]
You have to resolve the conflicts, then the CI will run (I think).

File: grammar/asdl_cpp.py [hunk 1]
```diff
@@ -405,6 +405,267 @@ def visitField(self, field):
...
+                    self.emit("self().visit_%s(x.m_%s[i]);" % (field.type, field.name), level+1)
+                self.emit(  'dec_indent();', level+1)      
+                self.emit("}", level)
+            elif field.opt: 
+                self.emit('s.append("\\n" + indtd + "%s" + "%s=");' % (arr, field.name), 2)
+                if last:
+                    self.emit('last = true;', 2)
+                    #self.emit('attached = true;', 2)
+                #else:
+                    #self.emit('attached = true;', 2)
```
Comment (namannimmo10) [inline grammar/asdl_cpp.py:546]:
```suggestion
```

File: grammar/asdl_cpp.py [hunk 2]
```diff
@@ -405,6 +405,267 @@ def visitField(self, field):
...
+                    self.emit(      's.append("\\n" + indtd + "|-counter=");', level)
+                    self.emit(      's.append(x.m_%s->get_counter());' % field.name, level)
+                    self.emit('size_t i = 0;', level)
+                    self.emit('s.append("\\n" + indtd + "└-scope=\u21a7");', level)
+                    self.emit('for (auto &a : x.m_%s->scope) {' % field.name, level)
+                    self.emit(      'i++;', level+1)
+                    self.emit(      'inc_indent();', level+1)
+                    self.emit(      'last = i == x.m_%s->scope.size();' % field.name, level+1)
+                    self.emit(      's.append("\\n" + indtd + (last ? "└-" : "|-") + a.first + ": ");', level+1)
+                    # self.emit(      'attached = true;', level+1)
```
Comment (namannimmo10) [inline grammar/asdl_cpp.py:622]:
```suggestion
```

**Comment by certik** [comment]
@ansharlubis I tested it locally and it works great! Thanks for this. Thanks @namannimmo10 for the fixes.

**Comment by ansharlubis** [comment]
@namannimmo10 thank you for fixing the mess :)

**Pull request merged**